### PR TITLE
Validate rules property

### PR DESCRIPTION
**Pull request by eddynaka**
Fixes #2127 

## Changes
- Validate if `driver.rules` is empty: `SARIF2012_ProvideRuleProperties_Note_ProvideMetadataForAllViolatedRules_Text`
- Validate if `result.ruleid` exists in `driver.rules`: `SARIF2012_ProvideRuleProperties_Note_ProvideRuleMetadata_Text`

cc: @marcogario

### CODE REVIEW

**Comment by marcogario** [comment]
Thanks for putting this together!

For my understanding, this rule now applies to all SARIF files, not only when using the GH Code Scanning feture, right?

**Comment by eddynaka** [comment]
> Thanks for putting this together!
> 
> For my understanding, this rule now applies to all SARIF files, not only when using the GH Code Scanning feture, right?

Hi @marcogario , yes, that would be applied for all sarif files as a **note**.

File: src/Test.FunctionalTests.Sarif/TestData/Multitool/ValidateCommand/ExpectedOutputs/SARIF2012.ProvideRuleProperties_Invalid.sarif [hunk 1]
```diff
@@ -17,14 +17,20 @@
...
               },
               "messageStrings": {
-                "Note_ProvideFriendlyName": {
-                  "text": "{0}: The rule '{1}' does not provide a \"friendly name\" in its 'name' property. The friendly name should be a single Pascal identifier, for example, 'ProvideRuleFriendlyName', that helps users see at a glance the purpose of the analysis rule."
-                },
                 "Note_FriendlyNameNotAPascalIdentifier": {
                   "text": "{0}: '{1}' is not a Pascal identifier. For uniformity of experience across all tools that produce SARIF, the friendly name should be a single Pascal identifier, for example, 'ProvideRuleFriendlyName'."
                 },
+                "Note_ProvideFriendlyName": {
+                  "text": "{0}: The rule '{1}' does not provide a \"friendly name\" in its 'name' property. The friendly name should be a single Pascal identifier, for example, 'ProvideRuleFriendlyName', that helps users see at a glance the purpose of the analysis rule."
```
Comment (michaelcfanning) [inline src/Test.FunctionalTests.Sarif/TestData/Multitool/ValidateCommand/ExpectedOutputs/SARIF2012.ProvideRuleProperties_Invalid.sarif:24]: >Pascal [](start = 145, length = 6)

Pascal-cased

**Review by michaelcfanning — CHANGES_REQUESTED** [review]
:clock1:

File: src/Sarif.Multitool.Library/Rules/RuleResources.Designer.cs [hunk 2]
```diff
@@ -822,9 +822,9 @@ internal class RuleResources {
...
         ///
-        ///Provide the &apos;name&apos; property, which contains a &quot;friendly name&quot; that helps users see at a glance the purpose of the rule. For uniformity of experience across all tools that produce SARIF, the friendly name should be a single Pascal identifier, for example, &apos;ProvideRuleFriendlyName&apos;.
+        ///Provide the &apos;name&apos; property, which contains a &quot;friendly name&quot; that helps users see at a glance the purpose of the rule. For uniformity of experience across all tools that produce SARIF, the friendly name should be a single Pascal-case identifier, for example, &apos;ProvideRuleFriendlyName&apos;.
         ///
-        ///Provide the &apos;helpUri&apos; property, which contains a URI where users can find detailed information about the rule. This information s [rest of string was truncated]&quot;;.
+        ///Provide the &apos;helpUri&apos; property, which contains a URI where users can find detailed information about the rule. This informat [rest of string was truncated]&quot;;.
```
Comment (michaelcfanning) [inline src/Sarif.Multitool.Library/Rules/RuleResources.Designer.cs:827]: >informat [](start = 137, length = 8)

Looks like you tried to remove that erroneous 's' but maybe deleted too much of the string?

File: src/Sarif.Multitool.Library/Rules/RuleResources.resx [hunk 3]
```diff
@@ -452,12 +452,18 @@ Semantics: Assuming the reader of the log file (an end user or another tool) has
...
   </data>
   <data name="SARIF2012_ProvideRuleProperties_Note_FriendlyNameNotAPascalIdentifier_Text" xml:space="preserve">
-    <value>{0}: '{1}' is not a Pascal identifier. For uniformity of experience across all tools that produce SARIF, the friendly name should be a single Pascal identifier, for example, 'ProvideRuleFriendlyName'.</value>
+    <value>{0}: '{1}' is not a Pascal-case identifier. For uniformity of experience across all tools that produce SARIF, the friendly name should be a single Pascal-case identifier, for example, 'ProvideRuleFriendlyName'.</value>
   </data>
   <data name="SARIF1004_ExpressUriBaseIdsCorrectly_Error_RelativeReferenceMustNotBeginWithSlash_Text" xml:space="preserve">
     <value>The relative reference '{0}' begins with a slash, which will prevent it from combining properly with the absolute URI specified by a 'uriBaseId'.</value>
   </data>
+  <data name="SARIF2012_ProvideRuleProperties_Note_ProvideMetadataForAllViolatedRules_Text" xml:space="preserve">
+    <value>'{0}' does not provide a 'rules' property. 'rules' contain information that helps users understand why each rule fires and what the user can do to fix it.</value>
```
Comment (michaelcfanning) [inline src/Sarif.Multitool.Library/Rules/RuleResources.resx:464]: >rules [](start = 37, length = 5)

Feel free to reach out to Larry if you'd like another set of eyes on this text proposal.

**Review by michaelcfanning — APPROVED** [review]
:shipit:

File: src/Sarif.Multitool.Library/Rules/RuleResources.Designer.cs (hunk 2 above)
Comment (eddynaka) [reply src/Sarif.Multitool.Library/Rules/RuleResources.Designer.cs:827]: This text is automatic generated when we change the resource itself. so, its breaking in a unusual space.

---
In reply to: [533635356](https://github.com/microsoft/sarif-sdk/pull/2161#discussion_r533635356) [](ancestors = 533635356)

File: src/Sarif.Multitool.Library/Rules/RuleResources.resx (hunk 3 above)
Comment (lgolding) [reply src/Sarif.Multitool.Library/Rules/RuleResources.resx:464]: LGTM

---
In reply to: [533635662](https://github.com/microsoft/sarif-sdk/pull/2161#discussion_r533635662) [](ancestors = 533635662)

**Pull request merged**